File: lab2-functionalprimer/python3/func_as_arg.py
# def listFunc():
#    return [i for i in range(1, 6)] #Create list of ints from 1 to 5, Haskell equivalent [1..5]
#
# def applicatorFunc(inpFunc, s):
#    if s=='s':
#        return sum(inpFunc())
#    else:
#        return sum(inpFunc())/5
#
# print(applicatorFunc(listFunc, 's'))

# Generate a list of integers from a to b
def inpFunc(a, b):
    return [i for i in range(a, b + 1)]  

# Take inpFunc as input and applies sum or average calculation
def applicatorFunc(numbers, s):
    if s == 's':
        return sum(numbers)  # Sum of the list
    else:
        return sum(numbers) / len(numbers)  # Average of the list

a = int(input("Enter the starting number (a): "))
b = int(input("Enter the ending number (b): "))
s = input("Enter 's' for sum or any other character for average: ")

numbers = inpFunc(a, b)
result = applicatorFunc(numbers, s)
print("Result:", result)

File: lab2-functionalprimer/haskell/func_as_arg.hs
{- --create inpFunc
inpFunc = [1..5] 

--Define applicatorFunc
applicatorFunc inpFunc s = if s=='s' then sum inpFunc else (sum inpFunc)/5  

main = doz
    let result = applicatorFunc inpFunc 'a' --Call applicatorFunc with inpFunc and 'a' as args
    putStrLn("sum = " ++ show(result))
-}

-- Takes two integers a and b and generates the list
inpFunc :: Int -> Int -> [Int]
inpFunc a b = [a..b]

-- applicatorFunc using guards
applicatorFunc :: [Int] -> Char -> Double
applicatorFunc nums s
    | s == 's'  = fromIntegral (sum nums)  -- Sum
    | otherwise = fromIntegral (sum nums) / fromIntegral (length nums) -- Average

main :: IO ()
main = do
    putStrLn "Enter the starting number (a):"
    a <- readLn
    putStrLn "Enter the ending number (b):"
    b <- readLn
    putStrLn "Enter 's' for sum or any other character for average:"
    s <- getChar
    let numbers = inpFunc a b
    let result = applicatorFunc numbers s
    putStrLn ("Result: " ++ show result)